### PR TITLE
Fix errors on Windows tests in new tests/functional

### PR DESCRIPTION
**Pull request by gshank**
resolves #4781

Fixing test failures in tests/functional/basic on Windows platforms.

### Description

Instead of generic 'write' methods, use text write methods with utf-8 encoding.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

**Comment by gshank** [comment]
Tests didn't fail. Also not failing on main now.

**Review by emmyoop — CHANGES_REQUESTED** [review]
One small request to document why we're doing this for future us.  Could also be worth starting a README with the reasons we do unexpected things.

File: core/dbt/tests/fixtures/project.py [hunk 1]
```diff
@@ -205,13 +208,15 @@ def write_project_files(project_root, dir_name, file_dict):
...
+            with open(path.join(name), "w", encoding="utf-8") as fp:
+                fp.write(value)
         elif name.endswith(".yml") or name.endswith(".yaml"):
             if isinstance(value, str):
                 data = value
             else:
                 data = yaml.safe_dump(value)
-            path.join(name).write(data)
+            with open(path.join(name), "w", encoding="utf-8") as fp:
+                fp.write(data)
```
[the remark this answers is not in the document]
Comment (gshank) [reply core/dbt/tests/fixtures/project.py]: Excellent idea, and much more readable. Have created 'read_file' and 'write_file' functions.

File: core/dbt/tests/util.py [hunk 2]
```diff
@@ -71,3 +71,17 @@ def copy_file(src_path, src, dest_path, dest) -> None:
 def rm_file(src_path, src) -> None:
     # remove files from proj_path
     os.remove(os.path.join(src_path, src))
+
+
+# We need to explicitly use encoding="utf-8" because otherwise on
+# Windows we'll get codepage 1252 and things might break
```
Comment (emmyoop) [inline core/dbt/tests/util.py:77]: 🎉

**Review by emmyoop — APPROVED** [review]

**Pull request merged**